Edited the Factories JSON Reference section for langua… (#624)
codenvy/che-docs: src/main/pages/che-6/portable-workspaces/factories_json_reference.adoc

folder: che-6/portable-workspaces
---

> A factory configuration is a JSON snippet either stored within {{ site.product_mini_name }} or as a `.factory.json` file. You can create factories within the IDE using the Che URL syntax, within the dashboard, or on the command line with the API and CLI.

Reviewer: No need to keep this variable around...

```suggestion
A factory configuration is a JSON snippet either stored within Che or as a `.factory.json` file. You can create factories within the IDE using the Che URL syntax, within the dashboard, or on the command line with the API and CLI.
```

[source,json]
----
factory : {
  "v"         : 4.0,            <1>
  "workspace" : {},             <2>
  "policies"  : {},             <3>
  "ide"       : {},             <4>
  "creator"   : {},             <5>
}
----
<1> Version of the configuration format.
<2> Identical to `workspace:{}` object for Eclipse Che.

> <3>(Optional) Restrictions that limit behaviors.

Reviewer: ```suggestion
<3> (Optional) Restrictions that limit behaviors.
```

<4> (Optional) Trigger IDE UI actions tied to workspace events.
<5> (Optional) Identifying information of author.

The `factory.workspace` is identical to the `workspace:{}` object for Eclipse Che and contains the structure of the workspace. To learn more about the workspace JSON object, see link:workspace-data-model.html[Workspace Data Model].

> You can export {{ site.product_mini_name }} workspaces and then reuse the workspace definition within a factory. {{ site.product_mini_name }} workspaces are composed of the following:

Reviewer: ```suggestion
You can export Che workspaces and then reuse the workspace definition within a factory. Che workspaces are composed of the following:
```

* 0..n projects
* 0..n environments that contain machines to run the code
* 0..n commands to execute against the code and machines
* a type

The `factory.policies`, `factory.ide`, and `factory.creator` objects are unique to factories. They provide meta information to the automation engine that alter the presentation of the factory URL or the behavior of the provisioning.

[id="mixins"]
== Mixins

A mixin adds additional behaviors to a project as a set of new project type attributes. Mixins are reusable across any project type. To define the mixins to add to a project, specify an array of strings, with each string containing the identifier for the mixin. For example, `"mixins" : [ "pullrequest" ]`.

> _Mixin ID_::pullrequest +

Reviewer: Not sure what the intention was here, but this isn't valid syntax. See how it renders.

>   _Description_: Enables pull request workflow where {{ site.product_mini_name }} handles local and remote branching, forking, and pull request issuance. Pull requests generated from within {{ site.product_mini_name }} have another factory placed into the comments of pull requests that a PR reviewer can consume. Adds contribution panel to the IDE. If this mixin is set, it uses attribute values for `project.attributes.local_branch` and `project.attributes.contribute_to_branch`. 

Reviewer: Again, why was the table removed?
```suggestion
  _Description_: Enables pull request workflow where Che handles local and remote branching, forking, and pull request issuance. Pull requests generated from within Che have another factory placed into the comments of pull requests that a PR reviewer can consume. Adds contribution panel to the IDE. If this mixin is set, it uses attribute values for `project.attributes.local_branch` and `project.attributes.contribute_to_branch`. 
```

> * The `pullrequest` mixin requires additional configuration from the `attributes` object of the project. If present, {{ site.product_mini_name }} will use the project attributes as defined in the factory. If not provided, {{ site.product_mini_name }} will set defaults for the attributes.

Reviewer: ```suggestion
* The `pullrequest` mixin requires additional configuration from the `attributes` object of the project. If present, Che uses the project attributes as defined in the factory. If not provided, Che uses set defaults for the attributes.
```

* Learn more about other link:TODO: link to project API docs[mixins]

[id="pull-request-mixin-attributes"]

> == pullrequest mixin attributes

Reviewer: Why?

Project attributes alter the behavior of the IDE or workspace.

> Different Eclipse Che and Codenvy plug-ins can add their own attributes to affect the behavior for the system. Attribute configuration is always optional and if not provided within a factory definition, the system will set it.

Reviewer 2: Behavior _for_ the system or _of_ the system? (What system are we even talking about here, by the way?)

The system will set attribute configuration to what? A system default?

Author: system = IDE or workspace. Edited it in the file.

> Different Eclipse Che and Codenvy plug-ins can add their own attributes to affect the behavior of the IDE or workspace. Attribute configuration is always optional and if not provided within a factory definition, the IDE or workspace will set it.

Reviewer: ```suggestion
Different Eclipse Che plug-ins can add their own attributes to affect the behavior of the IDE or workspace. Attribute configuration is always optional and if not provided within a factory definition, the IDE or workspace sets it.
```

[width="100%",cols="50%,50%",options="header",]
|===
|Attribute |Description

> |`local_branch` |Used in conjunction with the `pullrequest` mixin. If provided, the local branch for the project is set with this value. If not provided, the local branch is set with the value of `project.source.parameters.branch` (the name of the branch from the remote). If `local_branch` and `project.source.parameters.branch` are both not provided, the local branch is set to the name of the checked out branch.

Reviewer 2: "If both `local_branch` and `project.source.parameters.branch` are not provided"

|`contribute_to_branch` |Name of the branch that a pull request will be contributed to. The value of `project.source.parameters.branch` is default. It is the name of the branch that this project was cloned from.
|===

> Following is a snippet that demonstrates full configuration of the contribution mixin.

Reviewer 2: `contribution` should be styled, as it is a named mixin.

[id="policies"]
== Policies

Following is an example of a factory policy.

[source,json]
----
factory.policies : {
  "referer"   : STRING,                 <1>
  "since"     : EPOCHTIME,              <2>
  "until"     : EPOCHTIME,              <3>
  "create"    : [perClick | perUser]    <4>
}
----

> <1> Works only for clients from referer.

Reviewer: ```suggestion
<1> Works only for clients from a referrer.
```

<2> Factory works only after this date.
<3> Factory works only before this date.
<4> Create one workpace per click, user, or account.

[id="factories_json_reference-limitations"]
== Limitations

> You can use `since : EPOCHTIME`, `until : EPOCHTIME`, and `referer` as a way to prevent the factory from executing under certain conditions. `since` and `until` represent a valid time window that will allow the factory to activate. The `referer` will check the hostname of the acceptor and only allow the factory to execute if there is a match.

Reviewer: ```suggestion
You can use `since : EPOCHTIME`, `until : EPOCHTIME`, and `referer` as a way to prevent the factory from executing under certain conditions. `since` and `until` represent a valid time window that allows the factory to activate. The `referrer` checks the hostname of the acceptor and only allows the factory to execute if there is a match.
```

[id="factories_json_reference-multiplicity"]
== Multiplicity

> If `create : perClick` is used, then every click of the factory URL will generate a new workspace, each with its own identifier, name, and resources. If `create : perUser` is used, only one workspace is generated for each unique user that clicks on the factory URL. If the workspace has previously been generated, the existing workspace is reopened.

Reviewer: ```suggestion
Using `create : perClick` causes every click of the factory URL to generate a new workspace, each with its own identifier, name, and resources. Using `create : perUser` causes only one workspace to be generated for each unique user that clicks on the factory URL. If the workspace has previously been generated, the existing workspace is reopened.
```

[id="factories_json_reference-ide-customization"]

> == IDE customization

Reviewer: ```suggestion
== Customizing the IDE
```

[source,json]
----
factory.ide.{event} : {                 <1>
  "actions" : [{}]                      <2>
}

factory.ide.{event}.actions : [{
  "id"         : String,                <3>
  properties : {}                       <4>
}]
----
<1> event = `onAppLoaded`, `onProjectsLoaded`, `onAppClosed`.
<2> List of IDE actions to be executed when the event is triggered.

> <3> Action for the IDE to be performed when the event is triggered.

Reviewer 2: I'd recommend either "action for the IDE to perform" or "action to be performed by the IDE" here.

[width="100%",cols="34%,33%,33%",options="header",]
|===
|Action |Properties? |Description
|`warnOnClose` |No |Opens a warning pop-up window when the user closes the browser tab with a project that has uncommitted changes. Requires `project.parameters.keepVcs` to be `true`.
|===

[id="action-open-file"]
== Action: Open File

> This action opens a file as a tab in the editor. You can provide this action multiple times to have multiple files open. The file property is a relative reference to a file in the project source tree. The `file` parameter is the relative path within the workspace to the file that should be opened by the editor. The `line` parameter is optional and can be used to move the editor cursor to a specific line when the file is opened. Projects are located in the `/projects` directory of a workspace.

Reviewer: ```suggestion
This action opens a file as a tab in the editor. You can provide this action multiple times to have multiple files open. The file property is a relative reference to a file in the project source tree. The `file` parameter is the relative path within the workspace to the file that should be opened by the editor. The `line` parameter is optional and can be used to move the editor cursor to a specific line when the file is opened. Projects are located in the `/projects/` directory of a workspace.
```

> This action opens a file as a tab in the editor. You can provide this action multiple times to have multiple files open. The file property is a relative reference to a file in the project source tree. The `file` parameter is the relative path within the workspace to the file that should be opened by the editor. The `line` parameter is optional and can be used to move the editor cursor to a specific line when the file is opened. Note that projects are located in the workspaces `/projects` directory.

Reviewer 2: It looks like "workspaces `/projects` directory" is a typo which was originally meant to show that a workspace owns a `/projects` directory. Can that be rewritten to avoid the workspace showing possession? "Note that projects are located in the `/projects` directory of a workspace" (or something along those lines)?

Also, considering the use of the word "note", would this make sense as an admonition?

Author: Changed to "Projects are located in the `/projects` directory of a workspace." ( Got rid of "not that" because there was nothing noteworthy there.

[id="action-find-and-replace"]
== Action: Find and Replace

> If you create a project from a factory, you can have {{ site.product_mini_name }} find and replace values in the imported source code after it is imported into the project tree. This lets you parameterize your source code. Find and replace can be run as a *Run Command* during `onProjectsLoaded` event. You can use `sed`, `awk` or any other tools that are available in your workspace environment.

Reviewer: ```suggestion
In projects created from a factory, Che can find and replace values in the source code after it is imported into the project tree. This lets you parameterize your source code. Find and replace can be run as a *Run Command* during `onProjectsLoaded` event. You can use `sed`, `awk`, or any other tools that are available in your workspace environment.
```

}
----

> In the preceding example, a named command `replace` is created. The command replaces `***` with a string in project `README.md` file.

Reviewer 2: "The command replaces `***` with a string in the `README.md` file of the project."

Also, can we name the replacement string? "A string" is vague, while it clearly refers to "userId984hfy6" in the example `sed` command.

It would also be more accurate to say that it replaces _all_ occurrences of `***` (based on the use of `g` in `sed`).

> Then register this command to the execution list linked to the `onProjectsLoaded` event. In this example, `replace` command is executed after the project is imported into a workspace.

Reviewer 2: "the `replace` command"

  }
----

Use https://www.gnu.org/software/sed/manual/html_node/Regular-Expressions.html[regular expressions] in `sed`, both in find-replace and file-file type patterns.

[id="creator"]
== Creator

This object has meta information that you can embed within the factory. These attributes do not affect the automation behavior or the behavior of the generated workspace.

[source,json]
----
factory.creator : {
  "name"      : STRING,            <1>
  "email"     : STRING,            <2>
  "created"   : EPOCHTIME,         <3>
  "userId"    : STRING             <4>
}
----

> <1> Name of author of this configuration file.

Reviewer: ```suggestion
<1> The name of the author of this configuration file.
```
